### ISSUE TITLE
triggerCharacter not in options when completion is invoked manually

### ISSUE THREAD

**Issue by eliasnorrby** [neoclide/coc.nvim#3206]
## Result from CocInfo

```
vim version: NVIM v0.4.4
node version: v16.3.0
coc.nvim version: 0.0.80-2d813b6233
term: tmux
platform: linux
```

## Describe the bug

`triggerCharacter` is not part of the options object when completion is invoked manually.

When `Completion.startCompletion` is called via `Completion.triggerCompletion` (as it will be when `"suggest.autoTrigger": "always"`), the options object is enriched with the `triggerCharacter`:

```typescript
if (pre.length) {
  option.triggerCharacter = pre.slice(-1)
}
logger.debug('trigger completion with', option)
await this.startCompletion(option)
```
[context](https://github.com/neoclide/coc.nvim/blob/2d813b6233f7cb313e5c9d54ae2834348d982366/src/completion/index.ts#L409-L413)

But when completion is invoked manually (i.e. by setting up a binding for `coc#refresh()`), `Completion.startCompletion` is called with the raw options object:

```typescript
this.addAction('startCompletion', option => completion.startCompletion(option))
```
[context](https://github.com/neoclide/coc.nvim/blob/2d813b6233f7cb313e5c9d54ae2834348d982366/src/plugin.ts#L177)

I consider this a bug because `triggerCharacter` is not marked as optional in [`CompleteOption`](https://github.com/neoclide/coc.nvim/blob/2d813b6233f7cb313e5c9d54ae2834348d982366/src/types.ts#L424).

## Why is this an issue?

See neoclide/coc-git#185

When an extension (like `coc-git`) depends on a specific prefix (like '#' for completing GitHub issues), its completion will only work when triggered automatically.

## Reproduce the bug

0. (optional) Set `"suggest.autoTrigger": "none"`
1. Install `coc-git`: `:CocInstall coc-git`
2. Set up a binding for `coc#refresh()`

    ```vim
    inoremap <silent><expr> <TAB>
          \ pumvisible() ? "\<C-n>" :
          \ <SID>check_back_space() ? "\<TAB>" :
          \ coc#refresh()
    inoremap <expr><S-TAB> pumvisible() ? "\<C-p>" : "\<C-h>"
    ```

3. Open a commit message editor in a repo with issues (e.g. `git commit --allow-empty`)
4. Type '#' and press TAB

## A possible solution

In both cases, the base options object is produced by [`coc#util#get_complete_option`](https://github.com/neoclide/coc.nvim/blob/2d813b6233f7cb313e5c9d54ae2834348d982366/autoload/coc/util.vim#L338-L358). Note that it specifies all option attributes except `triggerCharacter`.

I suppose there's a reason `triggerCharacter` is set in TypeScript and not Vimscript (I learned today about vim's character and byte indexing.. problems). But maybe it's possible? How about:

```diff
function! coc#util#get_complete_option()
  let pos = getcurpos()
  let line = getline(pos[1])
  let input = matchstr(strpart(line, 0, pos[2] - 1), '\k*$')
+ let before = strpart(line, 0, pos[2] - 1)
+ let char = len(before) != 0 ? nr2char(strgetchar(before, strchars(before) - 1)) : ''
~ let input = matchstr(before, '\k*$')
  let col = pos[2] - strlen(input)
  let synname = synIDattr(synID(pos[1], col, 1), 'name')
  return {
        \ 'word': matchstr(strpart(line, col - 1), '^\k\+'),
        \ 'input': empty(input) ? '' : input,
        \ 'line': line,
+       \ 'triggerCharacter': char,
        \ 'filetype': &filetype,
        \ 'filepath': expand('%:p'),
        \ 'bufnr': bufnr('%'),
```

From what I can see, it handles both single-byte and double-byte characters properly.

## Another solution

Make `CompleteOption.triggerCharacter` optional and consider this a bug to be handled in `coc-git` instead.

**Comment by chemzqm** [comment]
Fixed by a56c2e80

**Comment by smackesey** [comment]
@chemzqm Your fix seems to have introduced a bug where I can no longer manually trigger completion if there is only whitespace before the cursor. I can trigger completion if I am in the first column, but not if I am in the second column with a space in front.
